src/vma/dev/cq_mgr.cpp

```diff
@@ -1600,3 +1599,356 @@ void cq_mgr::mlx5_init_cq()
+		*m_cq_dbell = htonl(m_cq_cons_index);
+		buff = m_rx_hot_buffer;
+		m_rx_hot_buffer = NULL;
+	} else {
+		prefetch((void*)m_rx_hot_buffer);
+	}
+
+	prefetch((void*)&(*m_cqes)[m_cq_cons_index & (m_cq_size - 1)]);
+
+	if (buff) {
+#ifdef RDTSC_MEASURE_RX_VERBS_READY_POLL
+		RDTSC_TAKE_END(RDTSC_FLOW_RX_VERBS_READY_POLL);
+#endif //RDTSC_MEASURE_RX_VERBS_READY_POLL
+
+#ifdef RDTSC_MEASURE_RX_READY_POLL_TO_LWIP
+		RDTSC_TAKE_START(RDTSC_FLOW_RX_READY_POLL_TO_LWIP);
+#endif
+	} else {
+#ifdef RDTSC_MEASURE_RX_VERBS_IDLE_POLL
+		RDTSC_TAKE_END(RDTSC_FLOW_RX_VERBS_IDLE_POLL);
+#endif
+
+#if defined(RDTSC_MEASURE_RX_VMA_TCP_IDLE_POLL) || defined(RDTSC_MEASURE_RX_CQE_RECEIVEFROM)
+		RDTSC_TAKE_START_VMA_IDLE_POLL_CQE_TO_RECVFROM(RDTSC_FLOW_RX_VMA_TCP_IDLE_POLL,
+			RDTSC_FLOW_RX_CQE_TO_RECEIVEFROM);
+#endif //RDTSC_MEASURE_RX_VMA_TCP_IDLE_POLL || RDTSC_MEASURE_RX_CQE_RECEIVEFROM
+	}
+
+	return buff;
+}
+
+inline void		cq_mgr_mlx5::cqe64_to_mem_buff_desc(volatile struct mlx5_cqe64 *cqe, mem_buf_desc_t* p_rx_wc_buf_desc, uint32_t& opcode, uint32_t& status)
+{
+	struct mlx5_err_cqe *ecqe;
+	ecqe = (struct mlx5_err_cqe *)cqe;
+
+	switch (cqe->op_own >> 4) {
+	case MLX5_CQE_RESP_WR_IMM:
+		cq_logerr("IBV_WC_RECV_RDMA_WITH_IMM is not supported");
+		opcode = IBV_WC_RECV_RDMA_WITH_IMM;
+		break;
+	case MLX5_CQE_RESP_SEND:
+	case MLX5_CQE_RESP_SEND_IMM:
+	case MLX5_CQE_RESP_SEND_INV:
+	{
+		opcode = VMA_IBV_WC_RECV;
+		p_rx_wc_buf_desc->sz_data = ntohl(cqe->byte_cnt);
+		status = IBV_WC_SUCCESS;
+
+#ifdef DEFINED_MLX5_HW_ETH_WQE_HEADER
+		//Checksum
+		p_rx_wc_buf_desc->rx.is_sw_csum_need = !(m_b_is_rx_hw_csum_on && (cqe->hds_ip_ext & MLX5_CQE_L4_OK) && (cqe->hds_ip_ext & MLX5_CQE_L3_OK));
+#else
+		p_rx_wc_buf_desc->rx.is_sw_csum_need = true;
+#endif
+
+		//Time stamp
+		p_rx_wc_buf_desc->rx.hw_raw_timestamp = cqe->timestamp;
+		//this is not a deadcode if timestamping is defined in verbs API
+		// coverity[dead_error_condition]
+	//	if (p_rx_wc_buf_desc->path.rx.poll_flags & VMA_IBV_WC_WITH_TIMESTAMP) {
+	//		p_rx_wc_buf_desc->path.rx.hw_raw_timestamp = 0;
+	//	}
+		return;
+	}
+	case MLX5_CQE_REQ://?!?!??!?!?!?!?!?!?!?!?
+		status = IBV_WC_SUCCESS;
+		return;
+	default:
+		status = IBV_WC_GENERAL_ERR;
+		break;
+	}
+
+	/* Only IBV_WC_WR_FLUSH_ERR is used in code */
+	if (MLX5_CQE_SYNDROME_WR_FLUSH_ERR == ecqe->syndrome) {
+		status = IBV_WC_WR_FLUSH_ERR;
+	} else {
+		status = IBV_WC_GENERAL_ERR;
+	}
+}
+
+int cq_mgr_mlx5::drain_and_proccess(uintptr_t* p_recycle_buffers_last_wr_id /*=NULL*/)
+{
+	cq_logfuncall("cq was %s drained. %d processed wce since last check. %d wce in m_rx_queue", (m_b_was_drained?"":"not "), m_n_wce_counter, m_rx_queue.size());
+
+	// CQ polling loop until max wce limit is reached for this interval or CQ is drained
+	uint32_t ret_total = 0;
+
+	if (p_recycle_buffers_last_wr_id != NULL) {
+		m_b_was_drained = false;
+	}
+
+	while ((m_n_sysvar_progress_engine_wce_max && (m_n_sysvar_progress_engine_wce_max > m_n_wce_counter)) &&
```

Comment:
`m_n_sysvar_progress_engine_wce_max` doesn't change during a VMA run so there's no need to retest it in a while condition. It's better to do:

```
        if (!m_n_sysvar_progress_engine_wce_max) {
                goto Exit;
        }

        while (m_n_sysvar_progress_engine_wce_max > m_n_wce_counter) {
                ...
        }

Exit: /* after the while (on line 1828 for example) */
```